pkg/logql/rangemapper.go

```diff
@@ -0,0 +1,313 @@
+	// and operation of the overrideDownstream expression in order to reduce the returned streams' label set.
+	if overrideDownstream != nil {
+		downstreamExpr = &syntax.VectorAggregationExpr{
+			Left:      downstreamExpr,
+			Grouping:  overrideDownstream.Grouping,
+			Operation: overrideDownstream.Operation,
+		}
+	}
+	return &syntax.BinOpExpr{
+		SampleExpr: &syntax.VectorAggregationExpr{
+			Left: m.mapConcatSampleExpr(downstreamExpr, rangeInterval),
+			Grouping: &syntax.Grouping{
+				Without: true,
+			},
+			Operation: syntax.OpTypeSum,
+		},
+		RHS:  &syntax.LiteralExpr{Val: rangeInterval.Seconds()},
+		Op:   syntax.OpTypeDiv,
+		Opts: &syntax.BinOpOptions{},
+	}
+}
+
+func (m RangeVectorMapper) vectorAggrWithRangeDownstreams(expr *syntax.RangeAggregationExpr, vectorAggrPushdown *syntax.VectorAggregationExpr, op string, rangeInterval time.Duration) syntax.SampleExpr {
+	grouping := expr.Grouping
+	if expr.Grouping == nil {
+		grouping = &syntax.Grouping{
+			Without: true,
+		}
+	}
+	var downstream syntax.SampleExpr = expr
+	if vectorAggrPushdown != nil {
+		downstream = vectorAggrPushdown
+	}
+	return &syntax.VectorAggregationExpr{
+		Left:      m.mapConcatSampleExpr(downstream, rangeInterval),
+		Grouping:  grouping,
+		Operation: op,
+	}
+}
+
+// appendDownstream adds expression expr with a range interval 'interval' and offset 'offset'  to the downstreams list.
+// Returns the updated downstream ConcatSampleExpr.
+func appendDownstream(downstreams *ConcatSampleExpr, expr syntax.SampleExpr, interval time.Duration, offset time.Duration) *ConcatSampleExpr {
+	sampleExpr, _ := clone(expr)
+	sampleExpr.Walk(func(e interface{}) {
+		switch concrete := e.(type) {
+		case *syntax.RangeAggregationExpr:
+			concrete.Left.Interval = interval
+			if offset != 0 {
+				concrete.Left.Offset += offset
+			}
+		}
+	})
+	downstreams = &ConcatSampleExpr{
+		DownstreamSampleExpr: DownstreamSampleExpr{
+			SampleExpr: sampleExpr,
+		},
+		next: downstreams,
+	}
+	return downstreams
+}
+
+// mapConcatSampleExpr transform expr in multiple downstream subexpressions split by offset range interval
+// rangeInterval should be greater than m.splitByInterval, otherwise the resultant expression
+// will have an unnecessary aggregation operation
+func (m RangeVectorMapper) mapConcatSampleExpr(expr syntax.SampleExpr, rangeInterval time.Duration) syntax.SampleExpr {
+	var downstreams *ConcatSampleExpr
+	interval := 0
+
+	splitCount := int(rangeInterval / m.splitByInterval)
+	for interval = 0; interval < splitCount; interval++ {
+		downstreams = appendDownstream(downstreams, expr, m.splitByInterval, time.Duration(interval)*m.splitByInterval)
+	}
+	// Add the remainder offset interval
+	if rangeInterval%m.splitByInterval != 0 {
+		offset := time.Duration(interval) * m.splitByInterval
+		downstreams = appendDownstream(downstreams, expr, rangeInterval-offset, offset)
+	}
+
+	if downstreams == nil {
+		return expr
+	}
```

Comment:
How about moving this up and use the `splitCount` instead?

```
if splitCount == 0 {
  return expr
}
```

Reply:
Good idea!